metadataserver/inject/inject_linux.go

```diff
@@ -0,0 +1,309 @@
+	if err != nil {
+		return nil, fmt.Errorf("Unable to move metadataservice interface from intermediate network namespace to container network namespace: %w", err)
+	}
+
+	err = intermediateNSHandle.LinkSetUp(tocontainer)
+	if err != nil {
+		return nil, fmt.Errorf("Could not set %q link up: %w", tocontainerInterfaceName, err)
+	}
+
+	metadataservice, err = containerNSHandle.LinkByName(toimdsInterfaceName)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get metadataservice link in container ns: %w", err)
+	}
+
+	err = containerNSHandle.LinkSetUp(metadataservice)
+	if err != nil {
+		return nil, fmt.Errorf("Could not set %q link up: %w", tocontainerInterfaceName, err)
+	}
+
+	// Make sure LO is up. This should be a noop:
+	lo, err := containerNSHandle.LinkByName("lo")
+	if err != nil {
+		return nil, fmt.Errorf("Could not get container lo: %w", err)
+	}
+	err = containerNSHandle.LinkSetUp(lo)
+	if err != nil {
+		return nil, fmt.Errorf("Could set container lo to up: %w", err)
+	}
+
+	// Add the IPs
+	err = intermediateNSHandle.AddrAdd(tocontainer, &netlink.Addr{
+		IPNet: v4ipnet,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Could not add %s to interface %q: %w", v4ipnet.String(), tocontainer.Attrs().Name, err)
+	}
+
+	err = intermediateNSHandle.AddrAdd(tocontainer, &netlink.Addr{
+		IPNet: v6ipnet,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Could not add %s to interface %q: %w", v6ipnet.String(), tocontainer.Attrs().Name, err)
+	}
+
+	var linkLocalAddr netlink.Addr
+	for ; err == nil; err = ctx.Err() {
+		addrs, err := intermediateNSHandle.AddrList(tocontainer, netlink.FAMILY_V6)
+		if err != nil {
+			return nil, fmt.Errorf("Could not list V6 addrs on %q: %w", tocontainer.Attrs().Name, err)
+		}
+
+		for _, addr := range addrs {
+			logger.G(ctx).Debugf("Checking addr %v, with flags %d", addr.IP, addr.Flags)
+			if addr.Scope == int(netlink.SCOPE_LINK) {
+				linkLocalAddr = addr
+				goto done
+			}
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+	return nil, fmt.Errorf("Could not get link local address on interface %q: %w", tocontainer.Attrs().Name, ctx.Err())
+done:
+
+	err = containerNSHandle.RouteAdd(&netlink.Route{
+		Via: &netlink.Via{
+			AddrFamily: netlink.FAMILY_V6,
+			Addr:       linkLocalAddr.IP,
+		},
+		Dst:       v4ipnet,
+		LinkIndex: metadataservice.Attrs().Index,
+		Src:       localhost,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Could not add route to %v: %w", v4ipnet, err)
+	}
+
+	err = containerNSHandle.RouteAdd(&netlink.Route{
+		Gw:        linkLocalAddr.IP,
+		Dst:       v6ipnet,
+		LinkIndex: metadataservice.Attrs().Index,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Could not add route to %v: %w", v6ipnet, err)
+	}
+
+	err = intermediateNSHandle.RouteAdd(&netlink.Route{
+		Dst:       localhostipnet,
+		LinkIndex: tocontainer.Attrs().Index,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Could not add route to %v: %w", localhostipnet, err)
+	}
+
+	runtime.LockOSThread()
```

Comment:
Can you add a comment as to why this is so far down here? Is it only because you have to call it *right* before `netns.Set` ?